classes/galley/DAO.inc.php

```diff
@@ -0,0 +1,308 @@
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class galley
+ *
+ * @brief Read and write galleys to the database.
+ */
+
+namespace PKP\galley;
+
+use APP\facades\Repo;
+use APP\publication\Publication;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\LazyCollection;
+use PKP\core\EntityDAO;
+use PKP\db\DAOResultFactory;
+use PKP\identity\Identity;
+use PKP\services\PKPSchemaService;
+use PKP\submission\PKPSubmission;
+use PKP\submission\Representation;
+use PKP\submission\RepresentationDAOInterface;
+
+class DAO extends EntityDAO implements RepresentationDAOInterface
+{
+    /** @copydoc EntityDAO::$schema */
+    public $schema = PKPSchemaService::SCHEMA_GALLEY;
+
+    /** @copydoc EntityDAO::$table */
+    public $table = 'publication_galleys';
+
+    /** @copydoc EntityDAO::$settingsTable */
+    public $settingsTable = 'publication_galley_settings';
+
+    /** @copydoc EntityDAO::$primarykeyColumn */
+    public $primaryKeyColumn = 'galley_id';
+
+    /** @copydoc EntityDAO::$primaryTableColumns */
+    public $primaryTableColumns = [
+        'submissionFileId' => 'submission_file_id',
+        'id' => 'galley_id',
+        'isApproved' => 'is_approved',
+        'locale' => 'locale',
+        'label' => 'label',
+        'publicationId' => 'publication_id',
+        'seq' => 'seq',
+        'urlPath' => 'url_path',
+        'urlRemote' => 'remote_url',
+        'doiId' => 'doi_id',
+    ];
+
+    public function newDataObject(): Galley
+    {
+        return app(Galley::class);
+    }
+
+    public function get(int $id): ?Galley
+    {
+        return parent::get($id);
+    }
+
+    public function getByUrlPath(string $urlPath, Publication $publication): ?Galley
+    {
+        $row = DB::table($this->table)
+            ->where('publication_id', $publication->getId())
+            ->where('url_path', $urlPath)
+            ->first();
+        return $row ? $this->fromRow($row) : null;
+    }
+
+    /**
+     * Get the number of galleys matching the configured query
+     */
+    public function getCount(Collector $query): int
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('g.' . $this->primaryKeyColumn)
+            ->get()
+            ->count();
+    }
+
+    /**
+     * Get a list of ids matching the configured query
+     */
+    public function getIds(Collector $query): Collection
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('g.' . $this->primaryKeyColumn)
+            ->pluck('g.' . $this->primaryKeyColumn);
+    }
+
+    /**
+     * Get a collection of galleys matching the configured query
+     */
+    public function getMany(Collector $query): LazyCollection
+    {
+        $rows = $query
+            ->getQueryBuilder()
+            ->select(['g.*'])
```

Comment:
I think the `select` should be left to the `Collector` to add right where the query is first created. That way downstream code can work with the select list, including using `addSelect` to add columns; if `select` is called later it clobbers the whole select list. (I've done this in `classes/user/Collector.inc.php` if you want an example.)

Reply:
In my view, the `getQueryBuilder()` method should not apply any `select()` statements, because it shouldn't presume to know what kind of information the caller wants to retrieve. You're right that the caller can override the `select()` with their own, but each calling method (`getIds()`, `getCount()`, and `getMany()`) will want to retrieve different things. 

If the `Collector::getQueryBuilder()` should default to selecting something, I think it should default to something more performant, like getting the IDs. Defaulting to the full row will encourage plugins to do things like this:

```php
$count = $collector
    ->getQueryBuilder()
    ->where(/* custom condition */)
    ->get()
    ->count();
```

Under-the-hood, that will instantiate `stdClass` objects for all row data just to perform a count on the `Collection`. By keeping the `select()` statement out of it, we can force plugins to choose their select statement, and therefore to think about the performance impacts of that choice.

I'm not wedded to this idea. It basically comes down to whether or not the `Collector` should bake in a default and, if so, what that default should be. But I do think that right now we're getting full objects too much, and this is costing us in performance. Maybe we'll solve this with caching in the future. But I think it's likely that we'll implement more DAO methods that fetch _more than_ ids/counts but _less than_ full entity objects. And in that sense, it might be good discipline for us to not have a default that presumes the query wants to retrieve everything.

Reply:
@asmecher in the interests of getting this merged, I split this discussion off to its own issue: https://github.com/pkp/pkp-lib/issues/7854.

I don't care too much which way we go. Let's just choose one, get it added to the docs, and apply it to all the `Collector`s.